### PR TITLE
Implement SPVM::List

### PR DESCRIPTION
**Pull request by motxx**
Implement SPVM::List: https://github.com/yuki-kimoto/SPVM/issues/16

## Interface
- Fixed at **2018/10/13**
```perl
sub new : SPVM::List ();
sub new_with_capacity : SPVM::List ($capacity : int);
sub new_with_array : SPVM::List ($objects : object[]);
sub length : int ($self : self);
sub push : void ($self : self, $value : object);
sub pop : object ($self : self);
sub unshift : void ($self : self, $value : object);
sub shift : object ($self : self);
sub set : void ($self : self, $index : int, $value : object);
sub get : object ($self : self, $index : int);
sub splice : object[] ($self : self, $cut_offset : int, $cut_length : int, $replace : object[]);
```
## Note
- Specification `splice[_with_list]` follows https://perldoc.perl.org/functions/splice.html as much as possible.
- autovivification is not supported.
- ~Throwing an exception in `SPVM::List` makes memory leak. If this specification is not good, please tell me what is best 🙇~ => It's default behavior. croak creates 2 objects, a header and a body.

### CODE REVIEW

**Comment by yuki-kimoto** [comment]
I see splice function.

  sub splice : SPVM::List ($self : self, $cut_offset : int, $cut_length : int);
  sub splice_with_list : SPVM::List ($self : self, $cut_offset : int, $cut_length : int, $insert_list : SPVM::List);

How about the following arguments and return value?

  sub splice : object[] ($self : self, $cut_offset : int, $cut_length : int, $replace : object[]);

**Comment by yuki-kimoto** [comment]
> For example, the following code throws exception and leaks 2 objects.

I will try the code after my current work. Maybe tomorrow morning.

**Comment by motxx** [comment]
> sub splice : object[] ($self : self, $cut_offset : int, $cut_length : int, $replace : object[]);

Yes, your interface will make the function simpler. Thank you.

File: lib/SPVM/List.spvm [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+    my $ret = $self->{values}[$self->{offset}];
+    $self->{values}[$self->{offset}] = undef;
+    $self->{offset}++;
+    return $ret;
+  }
+
+  sub set : void ($self : self, $index : int, $value : object) {
+    if (!$self->_range_check($index)) {
+      # do not allow autovivification
+      croak "Modification of non-creatable array value attempted, subscript " . $index;
```
Comment (motxx) [inline lib/SPVM/List.spvm]: Leak occurs. The following code doesn't leak.
```perl
warn ("Modification ...");
return;
```

**Comment by yuki-kimoto** [comment]
I'm reading SPVM::List now.

**Comment by yuki-kimoto** [comment]
I had told you a mistake about reallocation.

I realize that when reallocation, old values is copy to top of new values, offset become 0, it is simple.

Do you think this implementation?

**Comment by yuki-kimoto** [comment]
Do you think about capacity default  size?

    sub new : SPVM::List () {
      my $self = new SPVM::List;
      $self->{offset} = 0;
      $self->{length} = 0;
      
      # my $default_capacity = ?;
      # $self->{capacity} = $default_capacity;
      # $self->{values} = new object [$default_capacity];

      return $self;
    }

**Comment by yuki-kimoto** [comment]
I' reading t/default/lib-SPVM-List.t

**Comment by yuki-kimoto** [comment]
Very good test cases!

**Comment by yuki-kimoto** [comment]
I realized some points.

**Comment by motxx** [comment]
@yuki-kimoto
I added your new reviews without `benchmark`.
- [x] Un-support negative index for performance reasons.
- [x] Default capacity is introduced in `new` constructor.
- [x] `_reallocate` with `offset = 0`.
- [x] Remove throwing exceptions as much as possible for performance and simplicity.
  - Out-of-range exception is supported by SPVM compiler itself.
- [x] Use `if` and a bit duplicated code instead of using function calls.

**Comment by yuki-kimoto** [comment]
Thank you.

I'm reading commited codes now.

**Comment by motxx** [comment]
New commits to solve followings:
- [x] Fix overflow problems if a large index is used.
- [x] Improve splice algorithm in order not to always reallocate memory.
- [x] (Fix test) Separate testing contents of List from `check_fields` to `equals_list_deeply`. Tests are easier to write than before.

**Comment by yuki-kimoto** [comment]
moti. Thank you!

**Comment by yuki-kimoto** [comment]
Is it ok to merge? or need a little improvement?

**Pull request merged**